sky/server/requests/requests.py

```diff
@@ -864,57 +930,42 @@ async def get_api_request_ids_start_with(incomplete: str) -> List[str]:
-                              f'({", ".join(REQUEST_COLUMNS)}) VALUES '
-                              f'({", ".join(["?"] * len(REQUEST_COLUMNS))})')
-
-
-def _add_or_update_request_no_lock(request: Request):
-    """Add or update a REST request into the database."""
-    assert _DB is not None
-    with _DB.conn:
-        cursor = _DB.conn.cursor()
-        cursor.execute(_add_or_update_request_sql, request.to_row())
-
-
-async def _add_or_update_request_no_lock_async(request: Request):
-    """Async version of _add_or_update_request_no_lock."""
-    assert _DB is not None
-    await _DB.execute_and_commit_async(_add_or_update_request_sql,
-                                       request.to_row())
-
-
 def set_request_failed(request_id: str, e: BaseException) -> None:
     """Set a request to failed and populate the error message."""
+    assert request_id
     with ux_utils.enable_traceback():
         stacktrace = traceback.format_exc()
     setattr(e, 'stacktrace', stacktrace)
-    with update_request(request_id) as request_task:
-        assert request_task is not None, request_id
-        request_task.status = RequestStatus.FAILED
-        request_task.finished_at = time.time()
-        request_task.set_error(e)
+    req = update_request(request_id,
+                         set_status=RequestStatus.FAILED,
+                         set_finished_at=time.time(),
+                         set_error=e)
+    assert req is not None
 
 
 def set_request_succeeded(request_id: str, result: Optional[Any]) -> None:
     """Set a request to succeeded and populate the result."""
-    with update_request(request_id) as request_task:
-        assert request_task is not None, request_id
-        request_task.status = RequestStatus.SUCCEEDED
-        request_task.finished_at = time.time()
-        if result is not None:
-            request_task.set_return_value(result)
+    assert request_id
+    request = _get_request_no_lock(request_id)
+    assert request is not None
+    update_fields = ['status', 'finished_at']
+    request.status = RequestStatus.SUCCEEDED
+    request.finished_at = time.time()
+    if result is not None:
+        update_fields.append('return_value')
+        request.set_return_value(result)
+    _update_request(request, update_fields)
 
 
 def set_request_cancelled(request_id: str) -> None:
     """Set a pending or running request to cancelled."""
-    with update_request(request_id) as request_task:
-        assert request_task is not None, request_id
-        # Already finished or cancelled.
-        if request_task.status > RequestStatus.RUNNING:
-            return
-        request_task.finished_at = time.time()
-        request_task.status = RequestStatus.CANCELLED
+    assert request_id
+    req = update_request(
+        request_id,
+        set_status=RequestStatus.CANCELLED,
+        set_finished_at=time.time(),
+        match_status=[RequestStatus.PENDING, RequestStatus.RUNNING])
```

Comment:
Alternatively
```suggestion
        match_status=[status for status in RequestStatus if not (status > RequestStatus.RUNNING)])
```
if you want to be more robust against newly added `RequestStatus` values

Reply:
See https://github.com/skypilot-org/skypilot/pull/7661/commits/74c9be34746bc1bb050ae1fbb71f15bb9476adcb